heron/metricscachemgr/tests/java/com/twitter/heron/metricscachemgr/metricscache/CacheCoreTestHashIndex.java

```diff
@@ -0,0 +1,478 @@
+    request.componentNameInstanceId.put("c1", new HashSet<String>());
+    request.componentNameInstanceId.get("c1").add("i1");
+    request.componentNameInstanceId.get("c1").add("i2");
+    request.metricNames = new HashSet<String>();
+    request.metricNames.add("m1");
+    MetricsCacheQueryUtils.MetricResponse response = cacheCore.getMetrics(request, metricsFilter);
+    // there is 2 <component, instance, metric> tuples
+    Assert.assertEquals(response.metricList.size(), 2);
+    // sort
+    response.metricList.sort(comparator);
+    // there should be 2 instances
+    Assert.assertEquals(response.metricList.get(0).componentName, "c1");
+    Assert.assertEquals(response.metricList.get(0).instanceId, "i1");
+    Assert.assertEquals(response.metricList.get(0).metricName, "m1");
+    Assert.assertEquals(response.metricList.get(1).componentName, "c1");
+    Assert.assertEquals(response.metricList.get(1).instanceId, "i2");
+    Assert.assertEquals(response.metricList.get(1).metricName, "m1");
+    // there should be 1 metric for each instance
+    Assert.assertEquals(response.metricList.get(0).getMetricValue().size(), 1);
+    Assert.assertEquals(response.metricList.get(1).getMetricValue().size(), 1);
+    // check values
+    Assert.assertEquals(response.metricList.get(0).getMetricValue().get(0).value, "0.1");
+    Assert.assertEquals(response.metricList.get(1).getMetricValue().get(0).value, "0.3");
+  }
+
+  /*
+   * query instances: empty
+   */
+  @Test
+  public void testInstanceEmpty() {
+    MetricsCacheQueryUtils.MetricRequest request = new MetricsCacheQueryUtils.MetricRequest();
+    request.startTime = now - 95 * 1000;
+    request.endTime = now - 85 * 1000;
+    request.aggregationGranularity = RAW;
+    request.componentNameInstanceId = new HashMap<>();
+    request.componentNameInstanceId.put("c1", new HashSet<String>());
+    request.metricNames = new HashSet<String>();
+    request.metricNames.add("m1");
+    MetricsCacheQueryUtils.MetricResponse response = cacheCore.getMetrics(request, metricsFilter);
+    // there is 0 <component, instance, metric> tuples
+    Assert.assertEquals(response.metricList.size(), 0);
+  }
+
+  /*
+   * query components: null
+   */
+  @Test
+  public void testComponentNull() {
+    MetricsCacheQueryUtils.MetricRequest request = new MetricsCacheQueryUtils.MetricRequest();
+    request.startTime = now - 95 * 1000;
+    request.endTime = now - 85 * 1000;
+    request.aggregationGranularity = RAW;
+    request.metricNames = new HashSet<String>();
+    request.metricNames.add("m1");
+    MetricsCacheQueryUtils.MetricResponse response = cacheCore.getMetrics(request, metricsFilter);
+    // there is 4 <component, instance, metric> tuples
+    Assert.assertEquals(response.metricList.size(), 4);
+    // sort
+    response.metricList.sort(comparator);
+    // there should be 4 component-instance
+    Assert.assertEquals(response.metricList.get(0).componentName, "c1");
```

Comment:
This class has a ton of repeated logic that can be consolidated into test helpers to make the tests more readable and less verbose. For example with a helper like this you can assert each metricList item with one line:
```
import static org.junit.Assert.assertEquals;

private static void assertMetric(Metric metric, String expectedInstanceId, String expectedMetricName, String expectedComponentName, String... expectedValues) {
  assertEquals(metric.getComponentName(), expectedComponentName);
  assertEquals(metric.getInstanceId(), expectedInstanceId);
  assertEquals(metric.getMetricName(), expectedMetricName);
  assertEquals(metric.getMetricValue().size(), expectedValues.size();
  for (String expectedValue : expectedValues) {
    // assert values...
  }
} 
```